src/super_context.ml

```diff
@@ -245,8 +245,12 @@ let dump_env t ~dir =
   Ocaml_flags.dump (Env.ocaml_flags t ~dir)
 
 let resolve_program t ~dir ?hint ~loc bin =
-  let artifacts = Env.artifacts_host t ~dir in
-  Artifacts.binary ?hint ~loc artifacts bin
+  Build.arr
+    (fun args ->
+       let artifacts = Env.artifacts_host t ~dir in
+       let prog = Artifacts.binary ?hint ~loc artifacts bin in
```

Comment:
I don't think this is enough to get rid of the order dependence hack btw. Shouldn't we also change `Artifacts.binary` to return an arrow? Which would require changing `local_bins` to be `(unit, Path.t String.Map.t) Build.t Lazy.t`

Reply:
Yes, I started to do that but got stuck in `Super_context.Action.run`. I think that it might require changing some expander functions as well.

Reply:
Yup, that will be required. Since now we won't be able to expand artifacts statically anymore. We'll have to add something like we handle `Macro (Read, ..)`.